### PR TITLE
Improvements to ess_rhat

### PR DESCRIPTION
**Pull request by sethaxen**
This PR implements the nonbreaking parts of #22 (specifically, the proposal in https://github.com/TuringLang/MCMCDiagnosticTools.jl/issues/22#issuecomment-1338251834):
- makes splitting of chains optional
- support passing an `estimator` to compute an estimate-specific ESS (and R-hat too, though I don't think this is commonly used)
- Add `ess_rhat_bulk` to compute rank-normalized ESS and R-hat
- Add `ess_tail` and `rhat_tail` (the two cannot be simultaneously computed since they use different transformations)

~Currently this is marked as a draft while I write more tests and try to figure out how improve type-inferrability for utility functions.~

### CODE REVIEW

**Review by devmotion — COMMENTED** [review]
I haven't checked the PR in detail yet (since I'm on my phone and tests seem to fail) but one immediate suggestion is to avoid the JLD dependency and the JLD file. Either just use a text file or even better create the data in the tests without storing it.

**Comment by sethaxen** [comment]
> ...tests seem to fail

yes, the failures are in `rstar`, which is not touched here, so I'm guessing it's an issue with Random seed.

> ...one immediate suggestion is to avoid the JLD dependency and the JLD file. Either just use a text file or even better create the data in the tests without storing it.

Generating in the tests is cumbersome. When I tried sampling with Turing, I gave up after nearly an hour of sampling. StanSample can do it within a few minutes, but then we need CmdStan installed to run the tests locally and globally. Plus I wouldn't assume the tests would pass for all possible seeds to the sampler, so if CmdStan ever made a change to its sampler, our tests could suddenly begin failing. I'll add these draws as a text file.

**Comment by sethaxen** [comment]
> > ...tests seem to fail
> 
> yes, the failures are in `rstar`, which is not touched here, so I'm guessing it's an issue with Random seed.

The `rstar` tests are failing on the main branch, see #55

> ...one immediate suggestion is to avoid the JLD dependency and the JLD file. Either just use a text file or even better create the data in the tests without storing it.

Do you have suggestions of how to write a multidimensional array as a text file? I usually use JLD2 or NetCDF for this.

**Comment by devmotion** [comment]
Flatten it? I still think we should avoid binary files such as jld in the repo, both due to git and due to potential incompatibilities with future versions of Julia and/or JLD2 (similar to eg how JLD was abandoned and issues with BSON I ran into, even though probably they are less likely with simple arrays).

File: src/ess.jl
Comment (sethaxen) [inline src/ess.jl:345]: Since in the next breaking release, my goal is to split `ess_rhat` into `ess` and `rhat`, perhaps it's better to just declare `ess_bulk` and `rhat_bulk` here.

File: src/ess.jl
Comment (sethaxen) [inline src/ess.jl:226]:
```suggestion
```

**Review by sethaxen — COMMENTED** [review]
I found that by hand-coding the gradients I was able to get fast enough sampling to do it in the test suite. Now all tests pass except for those failing due to #56

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:232]: IMO an `ArgumentError` would be more appropriate:
```suggestion
    if x === nothing
        throw(ArgumentError("the estimator $f is not yet supported by `ess_rhat`"))
    end
```

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:233]: Couldn't we just use
```suggestion
    values = ess_rhat(Statistics.mean, x; kwargs...)
```
?

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:236]:
```suggestion
function ess_rhat(
    ::typeof(Statistics.mean),
```

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:351]: What's the reason for the promotion? Shouldn't it work fine without promotion?
```suggestion
```

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:353]:
```suggestion
        ess_rhat(Base.Fix2(Statistics.quantile, tail_prob / 2), x; kwargs...)[1],
```

File: src/ess.jl
Comment (devmotion) [inline src/ess.jl:354]:
```suggestion
        ess_rhat(Base.Fix2(Statistics.quantile, 1 - tail_prob / 2), x; kwargs...)[1],
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:25]: - Is it (supposed to be) exported? If not, maybe use `_split_chains` to make it clearer that it is an internal function?
- Do we need a default argument at all?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:29]: In particular if the fumction is not exported, it would be good to explain these things in the documentation of `ess_rhat`.

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:36]: Shouldn't this be the first index along the first axis only?
```suggestion
    data_sub = @views data[firstindex(data, 1):(end - niter_drop), :, :]
```
 Or on newer Julia versions simply
```suggestion
    data_sub = @views data[begin:(end - niter_drop), :, :]
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:121]: IMO we should use a different name. It's too similar to foldl and foldr in Base, and IMO not clear enough.

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:129]: Dispatch on `data` since according to the docstring `dims` is not supported for `data::AbstractVector`?

Moreover, as Julia does not specialize on functions, we want to use a type parameter here, I assume:
```suggestion
function fold(f::F, data::AbstractVector) where {F}
    return abs.(data .- f(data))
end
function fold(f::F data::AbstractArray{<:Any,3}; dims=(1,2)) where {F}
    return abs.(data .- f(data; dims=dims))
end
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:133]: Again, is this (supposed to be) exported? Otherwise maybe use `_rank_normalize`?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:122]: Do we need a default value?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:134]: Do we need a default value?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:138]: Would be good to explain this in the high-level exported functions where this is used.

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:156]: Without broadcasting?
```suggestion
    map!(StatsFuns.norminvcdf, values, values)
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:143]: I guess `map` + `eachslice` is not better?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:142]: Maybe dispatch on `dims` (see below)?
```suggestion
rank_normalize(x::AbstractArray{<:Any,3}; dims=(1, 2)) = _rank_normalize(x, dims)
function _rank_normalize(x::AbstractArray, dims)
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:145]: Could avoid `reshape` and `vec` if there is a generic implementation on `AbstractArray` without flattening:
```suggestion
        return _rank_normalize(xi, Colon())
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:148]:
```suggestion
rank_normalize(x::AbstractVector) = _rank_normalize(x, Colon())
function _rank_normalize(x::AbstractArray, ::Colon)
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:170]: Same as above, is this internal? If yes, maybe rename and add explanations to high-level API.

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:193]: Is this needed? Seems it could be avoided by using `mapslices`. Or even just set to `Bool`.
```suggestion
```

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:194]:
```suggestion
    y = similar(x, Bool)
```

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:351]: This is a workaround for https://github.com/JuliaStats/Statistics.jl/issues/136.

File: src/utils.jl
Comment (sethaxen) [reply src/utils.jl:25]: > * Is it (supposed to be) exported?

Originally I planned for it to be, but I no longer thing that is useful.

> * If not, maybe use `_split_chains` to make it clearer that it is an internal function?
> 
>     * Do we need a default argument at all?

`split_chain_indices` similarly has no leading underscore and has a default argument, and because the methods are related, I think it makes sense to keep them consistent in this regard.

File: src/utils.jl
Comment (sethaxen) [reply src/utils.jl:36]: `begin` isn't supported on Julia v1.3 (our oldest supported version). I didn't even know `firstindex` could take a dim argument.

File: src/utils.jl
Comment (sethaxen) [reply src/utils.jl:121]: Agreed. Originally we also folded about the mean (the old approach for computing ESS of `std`), but now we only use it for `median`, so `_fold_median` may be better.

**Comment by sethaxen** [comment]
@devmotion thanks for the detailed review. Since I now no longer intend for the user to directly call `fold` or `rank_normalize`, these no longer needed the `dims` keyword, which I have removed.

One other note: I removed definitions of R-hat as "potential scale reduction factor". While R-hat was originally motivated as estimating this quantity, it often fails to estimate it well but is still useful for diagnosing convergence e.g. from Vehtari et al:
image
Section 3 of https://arxiv.org/abs/2110.13017 give several other interpretations of R-hat.

Newer variants like bulk-Rhat, tail-Rhat, and their minimum are useful for diagnosing convergence but have no connection to PSRF. As a result, neither posterior nor ArviZ mention PSRF in their documentation of R-hat.

File: src/utils.jl
Comment (sethaxen) [reply src/utils.jl:25]: While BDA3 and Vehtari's paper do not discuss the case with odd chains, Stan, posterior, and ArviZ all discard the middle element. For `split_chains > 2`, we could generalize this by discarding the first `r` middle elements, where `r = rem(niter, split_chains)`. This requires bringing back `copy_split!`.

File: src/ess.jl
Comment (sethaxen) [inline src/ess.jl:232]: A flaw of this API is that it doesn't support estimators whose behavior is modified by keyword arguments. e.g. `StatsBase.mad` can take a user-provided `center` keyword, and we could trivially update `_expectand_proxy` to handle that, but there's no corresponding `Base.Fix` wrapper that handles such keywords. Will users need that? Probably not.

File: src/utils.jl
Comment (devmotion) [reply src/utils.jl:54]: Maybe `copyto!` with correct offsets etc. is safer?

File: src/utils.jl
Comment (devmotion) [inline src/utils.jl:154]:
```suggestion
    _rank_normalize(x::AbstractArray{<:Any,3})
```

File: test/ess.jl
Comment (devmotion) [inline test/ess.jl:63]: This is the default description/name anyway IIRC.
```suggestion
        @testset for scale in
```

File: test/ess.jl
Comment (devmotion) [inline test/ess.jl:161]: What's the purpose of `rng`? It seems you only call `ar1` with the default RNG, so maybe just remove the `rng` argument from `ar1` and sample with `randn(T, ...)` without RNG inside the function?

File: test/ess.jl
Comment (devmotion) [inline test/ess.jl:194]:
```suggestion
        x = 10 .+ mapreduce(hcat, σs) do σ
            return ar1(rng, φ, σ, ndraws, 1, nparams)
        end
```

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:298]: I noticed these as well. I'll remove `U` here. For `correctionfactor`, I'd prefer to leave it for another PR where I make sure there are no other places in our most common functions where we end up hardcoding a type like this.

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:361]: This definitely makes sense for Markdown. For docstrings and comments, I assume the reason for wrapping is so that it's easily readable on GitHub, where wrapping doesn't occur. (Markdown is just rendered as Markdown, so not an issue there). At any rate, BlueStyle wraps docstrings at 92 characters.

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:423]: Ah yes, this stopped being used when we dropped `_ess_mean`

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:11]: IIRC the reason I removed the duplicates was because we use `makedocs` with `strict=true`, which errors when there are duplicates. We'll need to change it to `strict=false`.

File: src/ess.jl
Comment (devmotion) [reply src/ess.jl:438]: Too bad.

File: src/ess.jl
Comment (sethaxen) [reply src/ess.jl:244]: Agreed, it's inelegant, but this is something we could support in the future. I'd probably wait until a user asks for it though.

**Comment by sethaxen** [comment]
I've implemented all requested changes. Only the windows tests fail, but that's expected. (see https://github.com/TuringLang/MCMCDiagnosticTools.jl/issues/56#issuecomment-1377306414)

The [references](https://turinglang.github.io/MCMCDiagnosticTools.jl/previews/PR53/) now look a little strange, but it seems to be the best we can do right now:
image

**Review by devmotion — APPROVED** [review]
Very nice, great PR! Looks good to me once the Linux test passes.

**Comment by sethaxen** [comment]
> Very nice, great PR! Looks good to me once the Linux test passes.

Thanks for the review! Can you merge? The repo settings require all tests pass before merge, and I don't have permissions to override.

**Comment by devmotion** [comment]
I changed the settings, now PRs can be merged (from everyone with merge permissions) even if Windows tests fail.

**Comment by sethaxen** [comment]
Thanks!

**Pull request merged**
